### ISSUE TITLE
Provide a better way to modify `CoroutineContext` in `RequestContext`

### ISSUE THREAD

**Issue by ks-yim** [line/armeria#4472]
Just an idea.

Currently, Armeria provides following ways to get & set `CoroutineContext` that will later be used to run a coroutine-based `AnnotatedService`:
* `CoroutineContexts#get`, `CoroutineContexts#set`
* `CoroutineContextService` with `CoroutineContextProvider`

...but I find `CoroutineContextService` and `CoroutineContextProvider` a bit redundant as what can be done with them can also be done with `CoroutineContexts#get/set` while the opposite is not true.

And non of those APIs provide a convenient way to combine(+) a new `CoroutineContext` with the one added earlier as `CoroutineContexts#set` just overwrites the whole `CoroutineContext` with the new one.

We may want to provide cleaner APIs to get/set `CoroutineContext` in a `RequestContext` by adding following extension functions & property, if there's no good reason to write those APIs in Java.

```kotlin
fun RequestContext.addCoroutineContext(ctx: CoroutineContext): Unit

fun RequestContext.removeCoroutineContext(ctx: CoroutineContext): Unit

var RequestContext.coroutineContext: CoroutineContext?
    get() = ...
    set(value) = ...

```

**Comment by minwoox** [comment]
IIRC, `CoroutineContextService` and `CoroutineContextProvider` are for adding additional user `CoroutineContext` before a Coroutine starts. 
https://github.com/line/armeria/blob/master/kotlin/src/main/kotlin/com/linecorp/armeria/internal/common/kotlin/ArmeriaCoroutineUtil.kt#L84
For example, if a user adds `CoroutineName` context, he/she can use it for debugging. https://github.com/line/armeria/blob/master/examples/annotated-http-service-kotlin/src/main/kotlin/example/armeria/server/annotated/kotlin/DecoratingService.kt#L53-L55
(Speaking of `CoroutineName`, I think we need a factory method in `CoroutineContextProvider` interface that provides `CoroutineName`.)
Meanwhile, `CoroutineContexts#get/set` are used to set the `CoroutineContext` in the decorator and retrieve it later. Having said that, I think they are different and we need them.

 > And non of those APIs provide a convenient way to combine(+) a new CoroutineContext with the one added earlier as CoroutineContexts#set just overwrites the whole CoroutineContext with the new one.
 
 That is true. We probably need the other method that adds/removes a `CoroutineContext`.
Just out of curiosity, could you give me examples of when the adding/removing methods should be used?

**Comment by ks-yim** [comment]
We use `coroutineContext` to store authN/Z context of an incoming request, but without using `CoroutineContextService` and `CoroutineContextProvider`.

We often just do the following in an `Authorizer<HttpRequest>` to add authenticated user context to the coroutine that will later be started while serving the request..:
```kotlin
class DefaultAuthorizer : Authorizer<HttpRequest> {
    override fun authorize(ctx: ServiceRequestContext, data: HttpRequest): CompletionStage<Boolean> {
        // ...verification
        CoroutineContexts.set(ctx, User(name = "name", role = "admin")
        return completedFuture(true)
    }
}

data class User(
    val name: String,
    val role: String
): AbstractCoroutineContextElement(User) {
    companion object Key : CoroutineContext.Key<User>
}
```

Yes, we do need `CoroutineContexts#get/set` as it is used in the above example, but I am suggesting the same functionality can be provided with more kotlin-friend APIs by leveraging kotlin's extension function & property (refer to the code written in my initial comment).

> Just out of curiosity, could you give me examples of when the adding/removing methods should be used?

Imagine the case where we want to add 2 different coroutine contexts from within 2 different decorator, though it doesn't seem like a right way to add a user-defined coroutine context that had been envisioned when `CoroutineContextService` and `CoroutineContextProvider` were introduced.

More specifically, let's say we want to store <ins>1) AuthN/Z context</ins> and <ins>2) DB's shard-routine context</ins> within 2 different decorators (note. This is a contrived example. I haven't found a practical use case of add/remove methods yet 😅).
```kotlin
fun buildServer(sb: ServerBuilder): Server {
    sb.decorator(ShardRoutingContextService.newDecorator())
      .decorator(AuthService.newDecorator(DefaultAuthorizer())
      .annotatedService(
          object {
              @Get("/foo/:id")
              suspend fun getFoo(@Param id: String) {
                  // Do whatever with the user-defined coroutine context elements.
              }
          }
    )
}
```

If we accidentally do `CoroutineContexts.set(ctx, shardRoutingCtx)` in `ShardRoutingContextService` in the above code, the authenticated user context that would have been added in `AuthService` will be overwritten and gone.

It's quite cumbersome that we should do...:
```kotlin
val currCoroCtx = CoroutineContexts.get(ctx);
CoroutineContexts.set(ctx, currCoroCtx + shardRoutingCtx);
```
...to make it correctly work.


Well, this wouldn't be an issue if I stick strictly to `CoroutineContextService` & `CoroutineContextProvider` approach, but then
1. we should always be careful of the decorator order
2. and the logic where the context elements are actually made should be located in different places(`respective decorator`) where the those context elements are added to the coroutine (`CoroutineContextService`).
... which I sometimes find annoying.

**Comment by minwoox** [comment]
> We often just do the following in an Authorizer<HttpRequest> to add authenticated user context to the coroutine 

Yeah, if there's another decorator that we can set a `CoroutineContext`, we don't need the `CoroutineContextService` decorator. But we do need `CoroutineContextService` if there's no decorator like this example: https://github.com/line/armeria/blob/master/examples/annotated-http-service-kotlin/src/main/kotlin/example/armeria/server/annotated/kotlin/DecoratingService.kt#L53-L55

 > but I am suggesting the same functionality can be provided with more kotlin-friend APIs by leveraging kotlin's extension function & property (refer to the code written in my initial comment).
 
 I agree with you on this. 👍 What I meant was that we need a way to set/add a `CoroutinContext` to `RequestContext` in a decorator. It doesn't have to be `CoroutineContexts#get/set` if there are better ones.
And the design looks nice. Are you sending a PR for that? 😆 
 
 > More specifically, let's say we want to store 1) AuthN/Z context and 2) DB's shard-routine context within 2 different decorators 
 
That's a possible scenario. 👍  

> this wouldn't be an issue if I stick strictly to CoroutineContextService & CoroutineContextProvider approach, but then

We didn't have to stick to `CoroutineContextService` and  `CoroutineContextProvider` from the first. 😆 
It's just a way to set the `CoroutineContext`. We can use another decorator and set `CoroutineContext` directly in it. 

/cc @okue

**Comment by ks-yim** [comment]
Thanks for inviting @okue, the original author. I'd definitely like to listen to his take on this.

**Comment by ikhoon** [comment]
`requestContext.addCoroutineContext(MDCContext())` looks to update the current `CoroutineContext` immediately, but the custom `CoroutineContext` is able to apply to the parent `CoroutineContext` of the `suspend` service just before it is called.
So I hesitated to add the method to the top level. To avoid such confusion, I prefer adding new functions to `CoroutineContexts`. 
For example:
```kt
CoroutineContexts.update(ctx, oldCtx -> oldCtx + newCtx);
CoroutineContexts.add(ctx, newCtx);
```

**Comment by ks-yim** [comment]
@ikhoon The approach you've suggested will prevent unnecessary confusion, thanks!
I will send a PR to add a new method to `CoroutineContexts`.
